### DISCUSSION TITLE
Direct x64 machine code generation - considerations for implementation for macOS and Windows

### DISCUSSION THREAD

**Discussion by ylluminate** [vlang/v#11680] [Ideas]
As we all know, x64 code generation is currently only available for Linux.  The purpose of this issue is to track, discuss and collaborate on efforts to get this also implemented for macOS and Windows (or to branch out into other issues to get more specific for each).

**Comment by sheerun** [comment]
Here are some resources for MacOS Mach-O format and assembler:

- [Writing 64-bit assembly on MacOS](http://www.idryman.org/blog/2014/12/02/writing-64-bit-assembly-on-mac-os-x/)
- [MacOS Assembler Reference](https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/Assembler/000-Introduction/introduction.html#//apple_ref/doc/uid/TP30000851-CH211-SW1)
- [Mach-O file format reference](https://github.com/aidansteele/osx-abi-macho-file-format-reference)

Something about statically compiled binaries:

https://developer.apple.com/library/archive/qa/qa1118/_index.html

> Apple does not support statically linked binaries on Mac OS X. A statically linked binary assumes binary compatibility at the kernel system call interface, and we do not make any guarantees on that front. Rather, we strive to ensure binary compatibility in each dynamically linked system library and framework.

> If your project absolutely must create a statically linked binary, you can get the Csu (C startup) module from Darwin and try building crt0.o for yourself. Obviously, we won't support such an endeavor.

Probably it also means that fully statically compiled binary won't be accepted to App Store (you cannot use direct syscalls), and is not backward or forward-compatible, so these are huge downsides. Also statically compiled crt0.o is not enough, [you also need statically compiled libSystem](https://github.com/skaht/Csu-85/issues/2#issuecomment-347373699) which simply isn't currently available. So probably it means we need  to live with it and figure out how to link system libraries (at least System) with V compiler.

So the proper way to compile assembly that uses libc calls (libSystem) to binary on MacOS is:

```
as main.s -o main.o && ld main.o -lc -o main
```

There are some interesting facts when it comes to linking when you read source of go-lang linker: https://golang.org/src/cmd/link/internal/ld/macho.go For example some offsets need to be maintained for code signing to be working

**Comment by ylluminate** [comment]
Due to Apple's stance on static bins, could parts of [Darling](https://www.darlinghq.org/) ([GitHub](https://github.com/darlinghq/darling) & [other associated repos](https://github.com/darlinghq?q=&type=&language=)) be used to help accomplish this?

**Comment by bjorn3** [comment]
Linux is one of the few systems that has a stable syscall interface. On all other systems you **must** use the system libraries to interface with the kernel. For macOS cross compilation there seems to be https://github.com/tpoechtrager/osxcross. For Windows cross compilation you can use mingw.

**Comment by medvednikov** [comment]
Exactly. V is going to be able to generate dynamically linked binaries.

For now it's Linux/static only.

**Comment by ylluminate** [comment]
For Windows @nord501 noted that it should be pe.v and asks whether or not we should use kernel32 or ntdll.lib...

**Comment by ylluminate** [comment]
@nord501 is putting forth a valiant and successful effort with Windows implementation and [noted this last night](https://discordapp.com/channels/592103645835821068/648394572501876746/723384231736246303) that he now has a working `hello world` and is digging into V's x64 gen for integration. Source: 
https://github.com/nord501/pe.v/blob/master/pe.v

**Comment by ylluminate** [comment]
macOS 11 (Big Sur) updates: https://gist.github.com/FiloSottile/7125822#gistcomment-3578065
